lib/WUI/nhttp/file_info.cpp

```diff
@@ -23,19 +23,116 @@ FileInfo::FileInfo(const char *filename, bool can_keep_alive, bool after_upload)
     : can_keep_alive(can_keep_alive)
     , after_upload(after_upload) {
     strlcpy(this->filename, filename, sizeof this->filename);
+    /*
+     * Eat the last slash on directories.
+     *
+     * This is mostly aestetic only... otherwise we get double slashs in the
+     * results/URLs. That works, but, meh...
+     */
+    const size_t len = strlen(this->filename);
+    if (len > 0 && this->filename[len - 1] == '/') {
+        this->filename[len - 1] = '\0';
+    }
 }
 
 Step FileInfo::step(std::string_view, bool, uint8_t *output, size_t output_size) {
-    // TODO: Support directories.
-
-    // We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
-    // We'll have to revisit this for directories, of course.
+    /*
+     * We assume the answer fits. If it doesn't, it doesn't overwrite anything, it just truncates the answer.
+     * Directories send each file as a single packet, but that one also assumes it fits.
+     */
     ConnectionHandling handling = can_keep_alive ? ConnectionHandling::ChunkedKeep : ConnectionHandling::Close;
     struct stat finfo;
-    if (stat(filename, &finfo) == 0) {
-        if (finfo.st_mode & S_IFDIR) {
-            return Step { 0, 0, StatusPage(Status::NotImplemented, can_keep_alive, "Directory listing not implemented") };
-        } else {
+    const size_t last_chunk_len = strlen(LAST_CHUNK);
+
+    if (dir) {
+        /*
+         * We already have a directory listing running. Try to pull another file out of there.
+         */
+        struct dirent *ent;
+        while ((ent = readdir(dir.get()))) {
+            // Check this is GCODE, not some other thing. Skip the others.
+            const char *last_dot = strrchr(ent->d_name, '.');
+            if (last_dot == nullptr) {
+                continue;
+            }
+            if (strcasecmp(last_dot, ".gcode") != 0 && strcasecmp(last_dot, ".gco") != 0) {
+                continue;
+            }
+
+            size_t written = render_chunk(handling, output, output_size, [&](uint8_t *output, size_t output_size) {
+                size_t written = 0;
+                if (need_comma) {
+                    assert(output_size > 0);
+                    output[0] = ',';
+                    written = 1;
+                } else {
+                    need_comma = true;
+                }
+                JSONIFY_STR(filename);
+                const char *name = ent->d_name;
+                JSONIFY_STR(name);
+#ifdef UNITTESTS
+                const char *long_name = ent->d_name;
+#else
+                const char *long_name = ent->lfn;
+#endif
+                JSONIFY_STR(long_name);
+                written += snprintf(reinterpret_cast<char *>(output + written), output_size - written,
+                    "{"
```

Comment:
What happens if the snprintf only manages to store a part of the whole string? Are we able to continue in the next packet? How?

Reply:
No, we are not able to continue in the next packet. In such case we would send a bogus answer (but not overwrite any memory or such). We assume packet buffers passed to these functions are „large enough“ (they may get split into smaller packets for the network transmission as necessary).

Being able to continue seems to be a somewhat hard problem that's not yet solved (and I'm not sure if it'll ever need systematic solving).

Nevertheless, it seems the usual answers are small enough and fit comfortably. I was actually thinking about optimizing and being able to put multiple files into the same buffer.

Reply:
ok